### PR TITLE
Add MLPerf UNet3D model

### PR DESCRIPTION
**Pull request by jla524**
Creating this draft PR for visibility.

Changes made:
- Add data loader for the KiTS19 dataset
- Implement functions to compute DICE score
- Implement functions for UNet3D inference

Download dataset:

```
git clone https://github.com/neheller/kits19
cd kits19
pip3 install -r requirements.txt
python3 -m starter_code.get_imaging
```

### CODE REVIEW

**Comment by geohot** [comment]
Cool! Will wait for e2e validation (confirming score is correct) before merging.

**Comment by geohot** [comment]
How is the score looking?

**Comment by jla524** [comment]
> How is the score looking?

I got a mean DICE score of 0.87 just now

**Comment by geohot** [comment]
There's merge conflicts right now, fix them and I'll test.

File: examples/mlperf/model_spec.py [hunk 1]
```diff
@@ -22,9 +22,10 @@ def spec_retinanet():
 
 def spec_unet3d():
   # 3D UNET
-  from models.unet3d import UNet3D
+  from models.unet3d_v2 import UNet3D
```
Comment (geohot) [inline examples/mlperf/model_spec.py:25]: Is this right?

File: models/unet3d.py [hunk 2]
```diff
@@ -1,42 +1,69 @@
...
-  def __init__(self, in_channels, out_channels):
-    self.groupnorm = nn.GroupNorm(1, in_channels) # 1 group?
-    # TODO: make 2D conv generic for 3D, might already work with kernel_size=(3,3,3)
-    self.conv = nn.Conv2d(in_channels, out_channels, kernel_size=(3,3,3), padding=(1,1,1,1,1,1), bias=False)
+import torch
+from tinygrad import nn
+from tinygrad.tensor import Tensor
+from extra.utils import download_file, get_child
+
+class ConvTranspose:
```
Comment (geohot) [inline models/unet3d.py:7]: There's one of these in nn now.

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -503,6 +503,10 @@ def layernorm(self, axis=-1, eps:float=1e-5) -> Tensor:
     y = (self - self.mean(axis, keepdim=True))
     return y.mul((y*y).mean(axis, keepdim=True).add(eps).rsqrt())
 
+  def instancenorm(self, axis=-1, eps:float=1e-5) -> Tensor: 
+    m, y = self.mean(axis, keepdim=True), (self*self).mean(axis, keepdim=True)
+    return (self - m) / ((y-m*m) + eps).sqrt()
```
Comment (geohot) [inline tinygrad/tensor.py:508]: Hmm, how is this different from layernorm?

File: examples/mlperf/model_spec.py (hunk 1 above)
Comment (jla524) [reply examples/mlperf/model_spec.py:25]: this should be just unet3d, I forgot to update it when I renamed the files

File: tinygrad/tensor.py (hunk 3 above)
Comment (jla524) [reply tinygrad/tensor.py]: looks like they are the same, I've removed this method

File: models/unet3d.py (hunk 2 above)
Comment (jla524) [reply models/unet3d.py]: updated

**Comment by geohot** [comment]
Hmm, so there's still an `nn.InstanceNorm` but tests for it were removed?

**Comment by geohot** [comment]
This is brutally slow...is it jitted? Why so much variance in model runtime?
```
5723.89 ms loading data, 18816.67 ms to run model
****** 0.89/1  0.89103 Mean DICE score
2717.41 ms loading data, 14307.90 ms to run model
****** 1.79/2  0.89584 Mean DICE score
7647.64 ms loading data, 30418.32 ms to run model
****** 2.39/3  0.79540 Mean DICE score
1772.79 ms loading data, 18041.00 ms to run model
****** 3.29/4  0.82160 Mean DICE score
1033.70 ms loading data, 13742.20 ms to run model
****** 4.18/5  0.83563 Mean DICE score
979.42 ms loading data, 7796.92 ms to run model
****** 5.12/6  0.85411 Mean DICE score
1309.51 ms loading data, 28314.90 ms to run model
****** 6.00/7  0.85709 Mean DICE score
574.63 ms loading data, 4817.27 ms to run model
```

**Comment by jla524** [comment]
> Hmm, so there's still an `nn.InstanceNorm` but tests for it were removed?

my bad, I put the tests back

**Comment by jla524** [comment]
> This is brutally slow...is it jitted? Why so much variance in model runtime?
> 
> ```
> tiny@tiny:~/tinygrad$ OPTLOCAL=2 MODEL=unet3d python3 examples/mlperf/model_eval.py
> eval unet3d
> /home/tiny/.local/lib/python3.10/site-packages/pyopencl/__init__.py:270: CompilerWarning: Non-empty compiler output encountered. Set the environment variable PYOPENCL_COMPILER_OUTPUT=1 to see more.
>   warn("Non-empty compiler output encountered. Set the "
> 5723.89 ms loading data, 18816.67 ms to run model
> ****** 0.89/1  0.89103 Mean DICE score
> 2717.41 ms loading data, 14307.90 ms to run model
> ****** 1.79/2  0.89584 Mean DICE score
> 7647.64 ms loading data, 30418.32 ms to run model
> ****** 2.39/3  0.79540 Mean DICE score
> 1772.79 ms loading data, 18041.00 ms to run model
> ****** 3.29/4  0.82160 Mean DICE score
> 1033.70 ms loading data, 13742.20 ms to run model
> ****** 4.18/5  0.83563 Mean DICE score
> 979.42 ms loading data, 7796.92 ms to run model
> ****** 5.12/6  0.85411 Mean DICE score
> 1309.51 ms loading data, 28314.90 ms to run model
> ****** 6.00/7  0.85709 Mean DICE score
> 574.63 ms loading data, 4817.27 ms to run model
> ```

Yes, it's jitted in `datasets/kits19.py` in line 95. It's much slower in torch, where the first case took at least 25 minutes to run on my m1 pro (cpu, since mps didn't support ConvTranspose2d).

I believe the variance in runtime is caused by the differences in input shapes. If you run `datasets/kits19.py` it will print the shapes of all images and labels in the eval set.
```
(1, 1, 190, 392, 392) (1, 190, 392, 392)                                                                                                                                               
(1, 1, 168, 365, 365) (1, 168, 365, 365)                                                                                                                                               
(1, 1, 260, 416, 416) (1, 260, 416, 416)                                                                                                                                               
(1, 1, 294, 316, 316) (1, 294, 316, 316)                                                                                                                                               
(1, 1, 278, 320, 320) (1, 278, 320, 320)                                                                                                                                               
(1, 1, 265, 283, 283) (1, 265, 283, 283)                                                                                                                                               
(1, 1, 343, 356, 356) (1, 343, 356, 356)                                                                                                                                               
(1, 1, 128, 289, 289) (1, 128, 289, 289) 
```

**Comment by geohot** [comment]
Cool, doing the final run now, if it hits the bounty is yours!

How do you want to be paid? Can either post here or e-mail me at george@tinygrad.org

**Comment by jla524** [comment]
> Cool, doing the final run now, if it hits the bounty is yours! It looks like it's just a lot of compute...
> 
> How do you want to be paid? Can either post here or e-mail me at [george@tinygrad.org](mailto:george@tinygrad.org)

Sounds good. My email is jackylee328@gmail.com.
